Lib/functools.py

```diff
@@ -192,6 +192,198 @@ def total_ordering(cls):
+        """
+        if self.ready_nodes is None:
+            raise ValueError("prepare() must be called first")
+        return self.nfinished < self.npassedout or bool(self.ready_nodes)
+
+
+    def done(self, node):
+        """Marks a nodes returned by "get_ready" as processed.
+
+        This method unblocks any successor of *node* for being returned in the future
+        by a call to "get_ready"
+
+        Raises :exec:`ValueError` if *node* has already been marked as processed by a
+        previous call to this method, if *node* was not added to the graph by using
+        "add" or if called without calling "prepare" previously or if node has not
+        yet been returned by "get_ready".
+        """
+
+        if self.ready_nodes is None:
+            raise ValueError("prepare() must be called first")
+
+        n2i = self.node2info
+
+        # Check if we know about this node (it was added previously using add()
+        if (nodeinfo := n2i.get(node)) is None:
+            raise ValueError(f"node {node!r} was not added using add()")
+
+        # If the node has not being returned (marked as ready) previously, inform the user.
+        stat = nodeinfo.npredecessors
+        if stat != _NODE_OUT:
+            if stat >= 0:
+                raise ValueError(f"node {node!r} was not passed out (still not ready)")
+            elif stat == _NODE_DONE:
+                raise ValueError(f"node {node!r} was already marked done")
+            else:
+                assert False, f"node {node!r}: unknown status {stat}"
+
+        # Mark the node as processed
+        nodeinfo.npredecessors = _NODE_DONE
+
+        # Go to all the successors and reduce the number of predecessors, collecting all the ones
+        # that are ready to be returned in the next get_ready() call.
+        for successor in nodeinfo.successors:
+            successor_info = n2i[successor]
+            successor_info.npredecessors -= 1
+            if successor_info.npredecessors == 0:
+                self.ready_nodes.append(successor)
+        self.nfinished += 1
+
+    def _find_cycle(self):
+        todo = set(self.node2info)
+
+        while todo:
+            node = todo.pop()
+            stack = [node]
+            # This set helps avoiding cuadratic behaviour when checking
+            # if a node is in the stack.
+            in_stack = {node}
+            while stack:
+                top = stack[-1]
+                for node in self.node2info[top].successors:
+                    if node in in_stack:
+                        return stack[stack.index(node):] + [node]
+                    if node in todo:
+                        in_stack.add(node)
+                        stack.append(node)
+                        todo.remove(node)
+                        break
+                else:
+                    node = stack.pop()
+                    in_stack.discard(node)
+        return None
```

Comment:
Ah - I see now that this algorithm works because it's willing to iterate over a node's successors over & over & over ... again.  Alas, while very clever, I don't think that can fly for production use.  For example, create an instance and then do

```python
for i in range(2, 100000):
    ts.add(i, 1)
```

There are no cycles, but there is a node with an enormous number of successors.  The cycle-finding algorithm in the `ts.py` I posted takes a fraction of a second (it's linear-time in all cases - which is _why_ it's hairier :wink:), but this one is still running after 5 minutes.

Reply:
> Ah - I see now that this algorithm works because it's willing to iterate over a node's successors over & over & over ... again. Alas, while very clever, I don't think that can fly for production use.

Good point! Although I liked the simplicity of this one, I agree that we should your proposal instead :)

> There are no cycles, but there is a node with an enormous number of successors. The cycle-finding algorithm in the ts.py I posted takes a fraction of a second (it's linear-time in all cases - which is why it's hairier wink), but this one is still running after 5 minutes.

Is the algorithm in `ts.py` a variation on Tarjan's algorithm?

Reply:
> Is the algorithm in `ts.py` a variation on Tarjan's algorithm?

Oh no.  I wish we had a blackboard here.  In words, it's a bog-standard depth-first search _solely_ looking for a cycle, but greatly obscured by using a manual stack to make it non-recursive.  But the only genuinely "hard part" (which isn't actually hard :wink:) is arranging to iterate over a node's successors exactly once each, which is done via a hand-maintained stack of successor iterators' `__next__` methods.

Reply:
Which _would_ be Tarjan's algorithm if you stripped all the recursion out of it, and also stripped out everything looking for strongly connected components - 🤣

Reply:
Cool! With a bit of a stretch (as is not the main algorithm here) and taking into account that we are *sorting* topologically, we could call this the second Tim sort in CPython ;)

Reply:
Added it in de2d2d9